src/transformers/tokenization_utils.py

```diff
@@ -408,20 +408,30 @@ def _add_tokens(self, new_tokens: Union[List[str], List[AddedToken]], special_to
         # Note: resize_token_embeddings expects to receive the full size of the new vocabulary, i.e. the length of the tokenizer.
         model.resize_token_embeddings(len(tokenizer))
         ```"""
-        new_tokens = [str(tok) for tok in new_tokens]
+        token_contents = [str(tok) for tok in new_tokens]
 
         tokens_to_add = []
-        for token in new_tokens:
+        for i, token in enumerate(token_contents):
             if not isinstance(token, str):
                 raise TypeError(f"Token {token} is not a string but a {type(token)}.")
             if not special_tokens and hasattr(self, "do_lower_case") and self.do_lower_case:
-                token = token.lower()
+                token = token.lower() # this is partly wrong. Added tokens that are not special but have escape cars are not
+                # lowered. Why means that the Trie will be wrong.
             if (
                 token != self.unk_token
                 and self.convert_tokens_to_ids(token) == self.convert_tokens_to_ids(self.unk_token)
                 and token not in tokens_to_add
             ):
                 tokens_to_add.append(token)
+                if isinstance(new_tokens[i], AddedToken) and not special_tokens :
+                    if hasattr(self, "do_lower_case") and self.do_lower_case:
+                        token = AddedToken(new_tokens[i].content.lower(), single_word = new_tokens[i].single_word, lstrip = new_tokens[i].lstrip, rstrip = new_tokens[i].rstrip, normalized = new_tokens[i].normalized)
+                    else:
```

Comment:
This highlights how bad the API is:
- you can't modify the content of the `AddedToken`
- if you only add the content to the unique no split: 
    - when decoding you will have a problem,since decoding uses `_additional_special_tokens`
    - when encoding, the `rstrip` and `lstrip` etc logic will be ignored 
- Adding a token does not necessarily update the `unique_no_split` etc